### PR TITLE
Feature/pending check

### PR DESCRIPTION
**Pull request by matthias-geuchen**
### Reason for this PR
- transactions requiring approvals should only be executable if the corresponding operation is still pending

### Changes in this PR
- check if an operation is pending before executing a transaction

### CODE REVIEW

**Review by anemonekk — APPROVED** [review]

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java [hunk 1]
```diff
@@ -123,24 +122,25 @@ public void validateApprovals(
...
+            operation = oUtil.getState(stub, key, OperationData.class);
         } catch (Exception e) {
-            approvals = new ApprovalList();
+            throw new ValidationError(GsonWrapper.toJson(getInsufficientApprovalsError()));
         }
         String clientId = getEnrollmentIdFromClientId(ctx.getClientIdentity().getId());
         List<String> clientGroups = new GroupContractUtil().getGroupNamesForUser(ctx.getStub(), clientId);
+        ApprovalList approvals = operation.getExistingApprovals();
         approvals.addUsersItem(clientId);
         approvals.addGroupsItems(clientGroups);
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java:141]: If no operation exists for a regular transaction, wouldn`t the code fail here?
I thought the default assumptions was: if no operation exists, we start at a new approvalList

Comment (NikoBergemann) [reply UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java:141]: i suggest something like:
```java
ApprovalList approvals;
State state;
try{
    OperationData oData = oUtil.getState(stub, key, OperationData.class);
    approvals = oData.getExistingApprovals();
    operation = oData.getState();
} catch (Exception e) {
    approvals = new ApprovalList();
    state = OperationState.Pending
}
```

Comment (matthias-geuchen) [reply UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java]: Good point! Updated it, but there is currently no transaction to test this with.

**Review by NikoBergemann — APPROVED** [review]

File: UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java [hunk 2]
```diff
@@ -123,24 +122,28 @@ public void validateApprovals(
...
             approvals = new ApprovalList();
+            operationState = OperationDataState.PENDING;
         }
         String clientId = getEnrollmentIdFromClientId(ctx.getClientIdentity().getId());
         List<String> clientGroups = new GroupContractUtil().getGroupNamesForUser(ctx.getStub(), clientId);
         approvals.addUsersItem(clientId);
         approvals.addGroupsItems(clientGroups);
 
-        if(!OperationContractUtil.covers(requiredApprovals, approvals)){
+        if(operationState != OperationDataState.PENDING || !OperationContractUtil.covers(requiredApprovals, approvals)){
```
Comment (NikoBergemann) [inline UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java:146]: should we do a different error?

Comment (matthias-geuchen) [reply UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/ContractUtil.java:146]: #138

**Pull request merged**